pippy/PipelineSchedule.py

```diff
@@ -669,42 +664,47 @@ def backward_stage_local_index(step):
         # Delay send waits
         sends_to_wait: List[dist.Work] = []
 
+        # Store ops (potentially across steps)
+        ops: List[dist.P2POp] = []
+
         for step in range(self.total_steps):
-            # warmup, forward only
+            # Warmup Phase (forward only)
             if step < warmup_steps:
-                logger.debug(f"{forward_stage_local_index(step)=}")
-
                 fwd_stage = self._stages[forward_stage_local_index(step)]
-                # assigns the current microbatch index and updates it for future steps
+
+                # This will assign the current microbatch index and update it for future steps
                 fwd_stage_mb_index[fwd_stage] = (
                     mb_index := fwd_stage_mb_index[fwd_stage]
                 ) + 1
 
                 logger.debug(
-                    f"{self.rank}: {step=}, {fwd_stage.stage_index=}, {mb_index=}"
+                    f"Rank {self.rank}: {step=}, {fwd_stage.stage_index=}, {mb_index=}"
                 )
 
                 with record_function(f"Forward {step}"):
-                    ops = fwd_stage.get_fwd_recv_ops()
-                    works = sorted_batch_isend_irecv(ops)
-                    for work in works.values():
+                    ops.extend(fwd_stage.get_fwd_recv_ops())
+                    if ops:
+                        work = dist.batch_isend_irecv(ops).pop()
                         work.wait()
+                        ops.clear()
 
                     output = fwd_stage.forward_one_chunk(arg_mbs[mb_index], kwarg_mbs[mb_index])  # type: ignore[index]
 
-                    ops = fwd_stage.get_fwd_send_ops()
-                    works = sorted_batch_isend_irecv(ops)
-                    sends_to_wait.extend(works.values())
+                    ops.extend(fwd_stage.get_fwd_send_ops())
+                    # If we are right before the fwd-bwd step, then we need to delay the send to the next step,
+                    # This is because fwd-bwd send/recvs among ranks need to be aligned to prevent a hang.
+                    # In the edge cases where there are no fwd_bwds and cooldown is immediate, then no delay is needed
+                    if ops and (step != warmup_steps - 1 or fwd_bwd_steps == 0):
+                        work = dist.batch_isend_irecv(ops).pop()
+                        sends_to_wait.append(work)
+                        ops.clear()
 
                     self._maybe_compute_loss(
                         fwd_stage, output, target_mbs, mb_index
                     )
 
-            # 1f1b
+            # 1F1B Phase (forward and backward)
             elif warmup_steps <= step < warmup_steps + fwd_bwd_steps:
```

Comment:
nit: `for step in range(warmup_steps, warmup_steps + fwd_bwd_steps)`
nit: rename `fwd_bwd_steps` to `1f1b_steps`